[Base] Patching System

Gliniak opened the pull request:
This is initial version of xex patching for Xenia.

It still requires some work, maybe a little bit of simplification and cleanup.

I created it as different project for future support for patching xex externally.

# Missing functionalities:
- CRC check or patching only when pattern matches
- Some future GUI functions?
- Saving changes to patch file directly from xenia (also for future GUI usage)

# Usage
1. There must be directory patches where xenia.exe is located (might be changed to Documents patch)
2. Patches for certain title **MUST** be  in file called <title_id>.patch (Example: 535107E4.patch)
3. To activate certain patch it must be enabled in .patch file and patching system must be enabled in xenia config
4. If title have any patch applied then in titlebar will be note ``[Patches Applied]``

Example patch:
```
title_name = "Gears Of War"
title_id = "4D5307D5"

[[patch]]
    name = "GoW - Texture rendering"
    desc = "None"
    author = "Gliniak"
    is_enabled = true
	
    [[patch.be32]]
        address = 0x822306D8
        value = 0x409AFF6C
```

# Concerns
- Having requirements of <title_id>.patch as patch filename might be problematic if there will be a lot of patches available. Title_id should be checked inside patch file (like in example) independently of its name. This will increase future readability and management.

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,213 @@
+
+PatchingSystem::~PatchingSystem() {}
+
+void PatchingSystem::loadPatches() {
```
UpperCase is used for method and class names (except for snake_case getters).

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,213 @@
+namespace patcher {
+
+PatchingSystem::PatchingSystem() {
+  isAnyPatchApplied_ = false;
```
snake_case_ is used for variables.

Triang3l commented on src/xenia/app/emulator_window.cc:
```diff
@@ -454,6 +454,12 @@ void EmulatorWindow::UpdateTitle() {
         ")";
   }
 
+  auto patching_system = emulator()->patching_system();
```
auto should generally be avoided because it makes code harder to read — only for cases when you already have a cast on the right side, or for things like iterators.

Triang3l commented on src/xenia/emulator.cc:
```diff
@@ -196,6 +196,8 @@ X_STATUS Emulator::Setup(
   // Bring up the virtual filesystem used by the kernel.
   file_system_ = std::make_unique<xe::vfs::VirtualFileSystem>();
 
+  patching_system_ = std::make_unique<xe::patcher::PatchingSystem>();
```
The word "system" is redundant, as most large classes are systems in general.

gibbed replied:
I would agree with Triang3l that using system in the name is a bit unclear, perhaps something like `PatchDB` would make more sense?

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,213 @@
+ * Released under the BSD license - see LICENSE in the root for more details. *
+ ******************************************************************************
+ */
+#include <regex>
```
Standard library includes are between the class header and other Xenia includes.

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,213 @@
+    alloc_size = sizeof(uint64_t);
+  }
+
+  if (provided_size == "be32") {
```
`else if`

Triang3l commented on src/xenia/patcher/patcher.h:
```diff
@@ -0,0 +1,80 @@
+      const std::string size_type,
+      const std::shared_ptr<cpptoml::table>& patch_table);
+
+  std::vector<patchFileEntry> getAllPatches() { return loaded_patch_files; }
```
This would create a copy of the vector — return a reference, `std::vector<PatchFileEntry>&`, instead.

Triang3l commented on src/xenia/patcher/patcher.h:
```diff
@@ -0,0 +1,80 @@
+
+}  // namespace patcher
+}  // namespace xe
+#endif
```
No new line at the end.

Gliniak commented:
Updated:
- Resolved most of review notes
- Removed dumb requirement of having `<title_id>.patch` filename template. Now you can use `<title_id><whatever>.patch`
(example: `4D5307D5.patch` can be named: `4D5307D5 - Gears Of War.patch`)

Triang3l commented on src/xenia/patcher/patcher.h:
```diff
@@ -0,0 +1,81 @@
+
+#include "xenia/memory.h"
+
+struct patchDataEntry {
```
Apart from incorrect case of the first letter, this should be in the xe::patcher namespace too.

Gliniak commented:
Updated:
 - Resolved review notes
 - Removed many "auto" usages
 - Added XXH3 Hash calculation (patch file might include generated hash, but it also might be skipped)

Margen67 commented on src/xenia/patcher/premake5.lua:
```diff
@@ -0,0 +1,14 @@
+  })
+  defines({
+  })
+  recursive_platform_files()
```
end of file newline

Gliniak commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  const std::filesystem::path patches_directory =
+      filesystem::GetExecutableFolder() / "patches";
```
Should we leave it as it is or maybe put it inside ``Documents\Xenia\Patches``?

gibbed replied:
This should be relative to user configuration in some fashion.

Triang3l replied:
Yes, relative to the storage root in our terms, which is the "home" directory of Xenia.

Triang3l replied:
Though maybe eventually we should support multiple search paths for patches.

gibbed commented on src/xenia/kernel/user_module.cc:
```diff
@@ -102,6 +115,8 @@ X_STATUS UserModule::LoadFromFile(const std::string_view path) {
     return result;
   }
 
+  XELOGI("Module hash: {:08X} for {}", hash_, name_);
```
`hash_` is a 64-bit value (uint64_t), so this should be `{:016X}`.

gibbed commented on src/xenia/kernel/user_module.h:
```diff
@@ -38,6 +38,7 @@ class UserModule : public XModule {
 
   const std::string& path() const override { return path_; }
   const std::string& name() const override { return name_; }
+  const uint64_t hash() const { return hash_; }
```
The return value doesn't need `const`-ness, just the method.

`uint64_t hash() const { return hash_; }`

gibbed commented on src/xenia/kernel/user_module.h:
```diff
@@ -100,6 +101,7 @@ class UserModule : public XModule {
 
   std::string name_;
   std::string path_;
+  uint64_t hash_;
```
Should initialize this to `0` or some other way to indicate that the value is invalid.

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+    if (!std::regex_match(patch_file.name.u8string(), file_name_regex_match)) {
+      XELOGE(
+          "PatchingSystem: Skipped loading file {} due to incorrect filename",
+          patch_file.name.u8string());
```
`to_utf8(...)`

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+
+PatchFileEntry PatchingSystem::ReadPatchFile(
+    const std::filesystem::path& file_path) {
+  PatchFileEntry patchFile;
```
Inconsistent variable naming. `file_entry`?

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+PatchFileEntry PatchingSystem::ReadPatchFile(
+    const std::filesystem::path& file_path) {
+  PatchFileEntry patchFile;
+  std::shared_ptr<cpptoml::table> patch_toml_fields;
```
Variable naming. `root_table` or `file_table`?

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+  patchFile.hash = strtoull((*title_hash).c_str(), NULL, 16);
+  patchFile.title_name = *title_name;
+
+  auto tarr = patch_toml_fields->get_table_array("patch");
```
Bad variable name. `patch_table_array`?

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+
+  auto tarr = patch_toml_fields->get_table_array("patch");
+
+  for (auto table : *tarr) {
```
If going with other proposed variable name changes. `table` should probably be `patch_table`?

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+
+    // Iterate through all available data sizes
+    for (const std::string& type : data_types) {
+      auto entries = ReadPatchData(type, table);
```
`patch_entries`

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+    return patch_data;
+  }
+
+  for (const auto& patch_data_table : *patch_data_tarr) {
```
It'd probably be better if this loop were in the caller of this function and this function explicitly only cared about parsing individual patch entries.

It'd probably be OK if there was another function that did the handling of iterating the `data_types` array, I just think the iteration should be the responsibility of the outer scope.

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+}
+
+uint8_t PatchingSystem::GetAllocSize(const std::string provided_size) {
+  uint8_t alloc_size = sizeof(uint32_t);
```
Needs better error handling, defaulting to `uint32_t` if an unknown size string is passed in is probably a bad idea.

gibbed commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,217 @@
+    if (patchFile.title_id != title_id) {
+      continue;
+    }
+    if (patchFile.hash != 0 && patchFile.hash != hash) {
```
Is it actually impossible for there to be a XXH3_64bits hash of `0`?

Margen67 commented on src/xenia/patcher/patch_db.cc:
```diff
@@ -0,0 +1,164 @@
+}
+
+}  // namespace patcher
+}  // namespace xe
```
eof newline

Margen67 commented on src/xenia/patcher/patch_db.h:
```diff
@@ -0,0 +1,78 @@
+}  // namespace patcher
+}  // namespace xe
+
+#endif  // XENIA_PATCH_LOADER_H_
```
eof newline

Margen67 commented:
> Updated:
> 
>     * Resolved most of review notes
> 
>     * Removed dumb requirement of having `<title_id>.patch` filename template. Now you can use `<title_id><whatever>.patch`
>       (example: `4D5307D5.patch` can be named: `4D5307D5 - Gears Of War.patch`)

This should be `.patch.toml` to be consistent with the config.

Margen67 commented on src/xenia/patcher/patch_db.cc:
```diff
@@ -0,0 +1,164 @@
+    uint32_t address = *patch_data_table->get_as<std::uint32_t>("address");
+    uint64_t value = *patch_data_table->get_as<std::uint64_t>("value");
+
+    if (size_type == "f32") {
+      float fvalue = float(*patch_data_table->get_as<double>("value"));
+      value = *(reinterpret_cast<uint32_t*>(&fvalue));
+    } else if (size_type == "f64") {
+      double dvalue = *patch_data_table->get_as<double>("value");
+      value = *(reinterpret_cast<uint64_t*>(&dvalue));
+    }
```
Maybe this should be a switch case?

Gliniak replied:
Sure, but I would need to add some string->enum map while I load each patch

Margen67 commented on src/xenia/patcher/patch_db.cc:
```diff
@@ -0,0 +1,164 @@
+  const std::vector<xe::filesystem::FileInfo>& patch_files =
+      filesystem::ListFiles(patches_directory);
+  const std::regex file_name_regex_match =
+      std::regex("^[A-Fa-f0-9]{8}.*\\.patch$");
+
+  for (const xe::filesystem::FileInfo& patch_file : patch_files) {
+    // Skip files that doesn't have only title_id as name and .patch as
```
Not sure if this is right...
```suggestion
      std::regex("^[A-F0-9]{8}.*\\.patch\\.toml$");

  for (const xe::filesystem::FileInfo& patch_file : patch_files) {
    // Skip files that doesn't have only title_id as name and .patch.toml as
```

Margen67 commented:
@Gliniak Title updates applied via xexp don't have the correct (updated) hash.

Triang3l commented on src/xenia/gpu/d3d12/premake5.lua:
```diff
@@ -39,6 +39,7 @@ project("xenia-gpu-d3d12-trace-viewer")
     "xenia-ui",
     "xenia-ui-d3d12",
     "xenia-vfs",
+    "xenia-patcher",
```
For easier maintenance, these lists should be sorted alphabetically unless the order matters (if it does, a comment explaining that is needed).

Triang3l commented on src/xenia/patcher/patch_db.cc:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  const std::filesystem::path patches_directory = patches_root_ / "patches";
+  const std::vector<xe::filesystem::FileInfo>& patch_files =
```
The usage of a reference here is confusing, will this vector object exist after the assignment, and also if ListFiles is called again afterwards?

Triang3l commented on src/xenia/patcher/patch_db.cc:
```diff
@@ -0,0 +1,223 @@
+  return title_patches;
+}
+
+void PatchDB::ReadHash(PatchFileEntry& patchEntry,
```
`patch_entry` *

Triang3l commented on src/xenia/kernel/user_module.h:
```diff
@@ -97,9 +98,11 @@ class UserModule : public XModule {
 
   std::string name_;
   std::string path_;
+  uint64_t hash_ = -1;
```
`UINT64_MAX` or `~UINT64_C(0)` as this is unsigned. Though neither this nor even 0 is an invalid hash value — if such indication is needed, `std::optional` should be preferred.

Triang3l commented on src/xenia/kernel/user_module.cc:
```diff
@@ -804,5 +808,39 @@ void UserModule::Dump() {
+      }
+      return page_index;
+    }
+    return (uint32_t)0;
```
The cast is not needed here. However, is 0 an invalid index (unlike something like `UINT32_MAX`) actually?

Triang3l commented on src/xenia/kernel/user_module.cc:
```diff
@@ -804,5 +808,39 @@ void UserModule::Dump() {
 }
 
+void UserModule::CalculateHash() {
+  const uint32_t page_size =
```
This should preferably be done via page size querying in `Memory`.

Triang3l commented on src/xenia/patcher/patch_db.h:
```diff
@@ -0,0 +1,131 @@
+  PatchDataValue(const size_t alloc_size, const T value)
+      : alloc_size_(alloc_size) {
+    patch_data_.resize(alloc_size_);
+    memcpy(patch_data_.data(), &value, alloc_size_);
```
`std::memcpy`, also `#include <cstring>` is needed.

Triang3l commented on src/xenia/patcher/patch_db.h:
```diff
@@ -0,0 +1,131 @@
+};
+
+enum class PatchDataType {
+  be8,
```
kBE8…kU16String.

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,67 @@
+Patcher::Patcher(const std::filesystem::path patches_root) {
+  is_any_patch_applied_ = false;
+  patch_db = new PatchDB(patches_root);
+}
```
`patch_db_` (with the underscore) *

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,67 @@
+  patch_db = new PatchDB(patches_root);
+}
+
+Patcher::~Patcher() {}
```
Unless there are issues with unique_ptr deletion, it should be either not declared at all, or defined as = default (though = default may be needed in the .cc rather than .h in some unique_ptr usage cases, it's somewhat confusing).

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,67 @@
+void Patcher::ApplyPatch(Memory* memory, const PatchInfoEntry* patch) {
+  for (const PatchDataEntry& patch_data_entry : patch->patch_data) {
+    uint32_t old_address_protect = 0;
+    auto address = memory->TranslateVirtual(patch_data_entry.memory_address_);
```
`auto` unless the type is obvious from the names in the assignment (or from an explicit cast) complicates reading the code.

Triang3l commented on src/xenia/patcher/patcher.cc:
```diff
@@ -0,0 +1,67 @@
+                  (uint32_t)patch_data_entry.new_data_.alloc_size_,
+                  kMemoryProtectRead | kMemoryProtectWrite);
+
+    memcpy(address, patch_data_entry.new_data_.patch_data_.data(),
```
`std::memcpy` and `#include <cstring>`.

Margen67 commented:
`title_id` should be removed since the title ID is part of the file name now.
